### ISSUE TITLE
Allow combination of Specification with different projections

### ISSUE THREAD

**Issue by gabrielheming** [ardalis/Specification#389]
Hello there,

I am using the Specification in a few projects that I am working on, and one of my projects came to the point that it has multiple Specifications only to represent different projections, using the same Where/Order expressions and/or Search Criteria.

This project in question is a WebAPI without any UI. At this moment, I am tweaking the current approach to map domain entities to their endpoints (VMs/DTOs). As some particular entities are loaded in different scenarios with partial data, we decided to map directly on the projection instead of loading the entity to memory and then mapping it to its VM/DTO.

My questioning came to the point that if I have to create a Specification per mapping/projection, is it still an advantage over the direct usage of DBContext? Of course, I am exaggerating it here, but it is only to make my point.

As Specification is easy to extend (extra points for you guys) I have created an extension method to combine a Specification with different projections.

My take on it is understanding whether it is useful/helpful or if it's just a load of BS making the purpose of Specification nonexistent.

In a nutshell, I created the `IProjectionSpecification` interface as below:

```
public interface IProjectionSpecification<T, TResult>
{
    public Expression<Func<T, TResult>>? Selector { get; }
    public void Select(Expression<Func<T, TResult>> selector);
}
```

And being able to use the extension method `.ToProjection`.

```
// VMs used as ex
public record StoreVm
{
    public int Id { get; set; }
    public string? Name { get; set; }
}

public record StoreWithStreetVm
{
    public int Id { get; set; }
    public string? Name { get; set; }
    public string? Street { get; set; }
}

// Store and StoreByIdSpec are concrete classes from Ardalis.Specification.UnitTests
ISpecification<Store> storeByIdSpec = new StoreByIdSpec(1);
ISpecification<Store, StoreVm> mySpecWithProjection = storeByIdSpec.ToProjection(new MyStoreToStoreVm());
ISpecification<Store, StoreWithStreetVm> mySpecWithProjection2 = storeByIdSpec.ToProjection(new MyStoreToStoreWithStreetVm());
```
I don't think it's necessary to go into implementation details right now, but I'm interested in hearing your thoughts.

If you believe it would be useful, I can submit it as a pull request or create an extension package.

**Comment by fiseni** [comment]
Hi @gabrielheming,

We already support projections in the specs, you have to inherit from `Specification<T, TResult>`.
If you want to have a single specification per entity, and reuse the same spec for arbitrary projections, then you may want to check our sample apps. More specifically the [sample3](https://github.com/ardalis/Specification/tree/main/sample/Ardalis.Sample.App3). In this sample, we're combining specs with Automapper for projections, and the usage becomes super simple. Of course, you can use other mapping libraries too.

```c#
// Projecting directly to response DTOs. In addition the response is paginated and wrapped in PagedResponse<T>.
app.MapGet("/customers", async (IReadRepository<Customer> repo,
                                [AsParameters] CustomerFilter filter,
                                CancellationToken cancellationToken) =>
{
    var spec = new CustomerSpec(filter);
    var result = await repo.ProjectToListAsync<CustomerDto>(spec, filter, cancellationToken);
    return Results.Ok(result);
});
```

**Comment by fiseni** [comment]
Btw, if you prefer doing the mapping manually, then you may create a spec extension for common criteria and create projection specification per result type. So, in this case, the projection specs will act as mappers (since you're doing it manually).

```c#
public static class CustomerSpecExtensions
{
    public static ISpecificationBuilder<Customer> ApplyCommon(this ISpecificationBuilder<Customer> builder)
    {
        // Apply your common queries for Customer here
        builder.Include(x => x.Addresses);
        return builder;
    }
}

public class CustomerDtoSpec : Specification<Customer, CustomerDto>
{
    public CustomerDtoSpec()
    {
        Query.ApplyCommon();
        Query.Select(x => new CustomerDto(
            x.Id, 
            x.Name, 
            x.Age, 
            x.Addresses.Select(a => new AddressDto(a.Id, a.Street, a.CustomerId)).ToList()));
    }
}
```

**Comment by gabrielheming** [comment]
Hi @fiseni, thank you for your feedback.

> We already support projections in the specs, you have to inherit from `Specification<T, TResult>`.

I thought it was clear that I was already using it, and it led to the creation of the `ToProjection` extension method solution to solve (through composition) the inheritance issue. However, I have to admit that I tend to go naturally to a composition approach when a virtual member is called in a constructor. On the other hand, mapping to a common projection with different criteria will lead to duplication. For instance: 2 different Specs but projects the same output.

My point is a similar approach to your example from sample3 without having to rely on an external library and using only the Specification library. That is the general question that I will rephrase: Do you think it might be useful for the community as a implementation or do you rather not have such implementation on the package?

Your example with `ApplyCommon` keeps the inheritance problem that sample3 solves.

On my take, I didn't want to add more methods on the already existing `IReadRepository<T>` and `IRepository<T>`. Though I will check the approach from sample3 to have a check on performance. I am testing part of my implementation using source generator to check any improvement, and I will see the results.

Thanks again for your feedback, it gave me more approaches to check on.

**Comment by fiseni** [comment]
@gabrielheming @ardalis 
Yes, I do understand your points. But, adding another construct `IProjectionSpecification` might not be the best way to move forward. We try to minimize the number of types that the library exposes. Also, we already have a construct `ISpecification<T, TResult>` that represents the same thing. So, we should find a way how to integrate that one into this usage. Roughly, I'm thinking in this direction

```c#
ISpecification<Customer> spec = new CustomerSpec();
ISpecification<Customer, CustomerDto> projectionSpec = new CustomerDtoSpec();

// Get a new spec by combining the state from the base spec.
var newSpec = projectionSpec.Combine(spec);
```

If we put it this way, then it's clear that we're trying to implement composite specifications. There have been thorough discussions on this topic, about the pros and cons of the composite specs. So, if we want to implement this feature then it should be done a bit more carefully and considering the broader scope, not just this specific case. Perhaps it's time we re-evaluate composite specs and the best way to move forward in that regard.

**Comment by gabrielheming** [comment]
@fiseni 

> we already have a construct `ISpecification<T, TResult>` that represents the same thing

> Perhaps it's time we re-evaluate composite specs and the best way to move forward in that regard.

True, and I agree with you.

On my scenario, I've used a new type to take advantage of source generator as it specifies that the direction will be from `ISpecification<T>` to `IProjectionSpecification<T, TResult>` resulting an `ISpecification<T, TResult>`, without generating code for something that is not going to be use. Though it could probably be done using an attribute and possibly some configuration.

Nevertheless, I would argue that the specification composition should be strict for the criteria, while projection an extension on top of the specification.

Furthermore, I agree with you that discussing composition in details could be the proper approach.

**Comment by fiseni** [comment]
Ok, we'll plan this. But, before starting this work, we need substantial refactoring of the internals. It's something I've been planning for quite some time. Only then, we can start implementing this feature.

**Comment by fiseni** [comment]
I want to revisit this topic. Let's first define the goals.
- We want to be able to combine a given specification with the projection defined in another specification.
- This offers the ability to use specs as containers for mapping. So, we can "attach" this mapping to any specification.
- We should create a new specification instead of mutating the state of the input specs.

I'm thinking in this direction. Let's say we have the following constructs.

```csharp
public class Customer
{
    public int Id { get; set; }
    public string? Name { get; set; }
    public string? Address { get; set; }
}

public class CustomerDto
{
    public int Id { get; set; }
    public string? Name { get; set; }
}

public class CustomerSpec1 : Specification<Customer>
{
    public CustomerSpec1(string name)
    {
        Query.Where(x => x.Name == name);
    }
}

public class CustomerSpec2 : Specification<Customer>
{
    public CustomerSpec2(int id)
    {
        Query.Where(x => x.Id == id);
    }
}

public class CustomerDtoSpec : Specification<Customer, CustomerDto>
{
    public CustomerDtoSpec()
    {
        Query.Select(x => new CustomerDto
            {
                Id = x.Id,
                Name = x.Name
            });
    }
}
```

Now, we may combine them as follows.

```csharp
var customerSpec1 = new CustomerSpec1("John");
var customerSpec2 = new CustomerSpec2(1);
var customerDtoSpec = new CustomerDtoSpec();

// Creates new `Specification<Customer, CustomerDto>` spec,
// with the state of CustomerSpec and only the projection of CustomerDtoSpec.
// The input specs remain unchanged.
var newSpec1 = customerSpec1.WithProjectionOf(customerDtoSpec);
var newSpec2 = customerSpec2.WithProjectionOf(customerDtoSpec);
```

I chose the method name `WithProjectionOf` to make it clear that it's the only state combined. The projection state may contain additional expressions, but that's ignored.

@gabrielheming @ardalis Any thoughts? I'm open to ideas.

**Comment by gabrielheming** [comment]
I agree with the goals and also the implementation direction. I believe it keep at zero any potential breaking changes, and it will make possible the combination between specifications.

In that sense, I believe I have most of the implementation for it is already done on my side. It would be mostly a matter of adapting to the version 9 and keep it within the goals.

**Comment by fiseni** [comment]
Thanks @gabrielheming for the inputs. The feature is implemented.

**Note:**
We're creating a copy of the base state only. The user specification may contain additional custom states that are not carried out. We'll consider a `MemberwiseClone` in the future (I'm still undecided). Also, expressions may capture the user specification state, so the copy will reference the source spec indirectly. We'll see how users will utilize this feature and will make further adjustments based on the feedback.

**Comment by fiseni** [comment]
This feature is deployed in version 9.1.0
